src/sync/barrier.rs

```diff
@@ -56,57 +67,77 @@ impl Barrier {
+            trace!(waiters=?state.waiters, epoch=my_epoch, "blocked on barrier {:?}", self);
             ExecutionState::with(|s| s.current_mut().block(false));
-            trace!(leader=?state.leader, waiters=?state.waiters, "blocked on barrier {:?}", self);
         } else {
-            trace!(leader=?state.leader, waiters=?state.waiters, "releasing waiters on barrier {:?}", self);
-            // Unblock each waiter, updating its clock with the barrier's clock
+            trace!(waiters=?state.waiters, epoch=my_epoch, "releasing waiters on barrier {:?}", self);
+
+            debug_assert!(state.waiters.len() == state.bound || state.bound == 0);
+
+            // Make the leader token available for this epoch. The first task to wake up will
+            // take it and become the leader. The token shouldn't already be available.
+            assert!(state.leader_tokens.insert(my_epoch));
+
+            // Drain the set of waiters and increment the barrier's epoch, so any other task that
+            // calls `wait` from now on becomes part of a separate group with its own leader.
+            let waiters: Vec<TaskId> = state.waiters.drain().collect();
+            state.epoch += 1;
+
+            trace!(
+                waiters=?state.waiters,
+                epoch=state.epoch,
+                "releasing waiters on barrier {:?}",
+                self,
+            );
+
             let clock = state.clock.clone();
             ExecutionState::with(|s| {
-                for tid in state.waiters.drain() {
-                    debug_assert_ne!(tid, me);
+                // `waiters` includes the current task.
+                for tid in waiters {
                     let t = s.get_mut(tid);
-                    debug_assert!(t.blocked());
                     t.clock.increment(tid);
                     t.clock.update(&clock);
-                    t.unblock();
+
+                    // A thread blocked on the barrier will be blocked iff it's not the current
+                    // one, since this is the last thread arriving at the barrier for this epoch.
+                    assert_eq!(t.blocked(), tid != me);
+                    if t.blocked() {
+                        t.unblock();
+                    }
                 }
-                // Update the leader's clock
-                let t = s.current_mut();
-                t.clock.increment(me);
-                t.clock.update(&clock);
             });
-        }
-
-        let result = BarrierWaitResult {
-            is_leader: state.leader.unwrap() == me,
         };
 
         drop(state);
 
+        // Mark the task as requesting an explicit yield and schedule a thread switch, otherwise
+        // some schedulers would nearly always pick to continue executing the current task.
+        // Requesting a yield allows other threads to be the first one to wake up and grab the
+        // leader token.
+        ExecutionState::request_yield();
```

Comment:
Unlike #101 I think we don't want to yield in this case. We needed it there for fairness because `park` gets used in spinloops, which are doomed by unfair schedules. But barriers aren't used like that—threads generally do useful work once unblocked by a barrier. (I do wish we had better rules for thinking about this, but it's all random in the end anyway, so it gets a bit handwave-y).

Reply:
My intention here was to prevent an unfair schedule from making it so that it's (almost) always the last thread that gets to the barrier that becomes the leader, by giving every other thread a chance to wake up first and become the leader instead. In most situations I can think of though, it's probably already random enough which thread happens to call wait last, so the "extra" fairness here won't help much. I'll remove the `request_yield` here.